packages/react/src/components/NumberInput/NumberInput.js

```diff
@@ -38,7 +40,11 @@ const capMax = (max, value) =>
 class NumberInput extends Component {
   constructor(props) {
     super(props);
-    let value = props.value;
+    let value =
+      useControlledStateWithValue || props.value === undefined
```

Comment:
**Important: The `defaultValue` should be completely ignored if a value prop is passed in.** It's only purpose is to supply an initial state for an otherwise uncontrolled component. With your current implementation, the `defaultValue` will override a `value` prop if both were passed in.

A few more things:

* I think we can clean up this this logic with the `capMin` helper
* We should refactor to avoid relying on undefined checks and avoid setting variables to undefined. 
* This refactoring means we can set a proper defaultProp for the `devaultValue` (0).
* We should also add a `controlledMode` instance field to consolidates the flag checks. In my opinion, an instance field is more appropriate than state since we never intend on changing the property during the life cycle of the component.

```js
  constructor(props) {
    super(props);
    const { value: valueProp, defaultValue, min } = props;
    this.controlledMode =
      useControlledStateWithValue && valueProp !== undefined;

    const value = this.controlledMode ? valueProp : defaultValue;
    this.state = { value: capMin(min, value) };
  }
  
  // Add to defaultProps
  // static defaultProps = {
  //   defaultValue: 0
  // };
```

I have a sandbox with [a demo of the constructor](https://codesandbox.io/s/festive-perlman-qgc2w)

@emyarod this behavior might be related to #3013